Block: Outline when interacting with Toolbar Block Type/Movers

ItsJonQ opened the pull request:
## Description

<img width="751" alt="Screen Shot 2020-03-16 at 3 58 55 PM" src="https://user-images.githubusercontent.com/2322354/76795369-60d11e80-679f-11ea-9c0f-f248a7e81401.png">

This update enhances the new G2-based Toolbar interactions to provide a highlighted outline for a block when interacting with the Block Type/Mover actions.

## How has this been tested?

* Tested locally in Gutenberg

## Screenshots

![Screen Capture on 2020-03-16 at 15-49-00](https://user-images.githubusercontent.com/2322354/76795441-8bbb7280-679f-11ea-8870-b3d91cd94d51.gif)

The GIF (above) demonstrates the outline when engaging with the Block Type/Mover actions.

### Interaction

The outline shows when:

* Block Type is focused or hovered
* Any mover is focused or hovered

The style of the highlight is the **same** as the block's navigation mode styles. This can be seen by **pressing ESC** when on a block.

### Top Toolbar

<img width="800" alt="Screen Shot 2020-03-16 at 4 13 44 PM" src="https://user-images.githubusercontent.com/2322354/76796125-24062700-67a1-11ea-9733-138b2f0d7603.png">


The highlight feature still works for Top Toolbar (as well as smaller viewports)


## Types of changes
* Adding new `blockHighlighted` state to track the highlighted block, responding to Block Toolbar interactions
* Added reducer, action, and selector to accompany the `blockHighlighted` state
* Hook up the action with `BlockToolbar`
* Hook up state to render with `block-list/block`

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Resolves: https://github.com/WordPress/gutenberg/issues/20875

Commit:
Add action, reducer, and selector for new blockFocused state

Commit:
Render highlight UI for block when interacting with Block toolbar

mtias commented:
This feels very good, thanks for putting it together. A few notes:

- Let's try without the debounce for the highlight — it adds a delay which makes it hard to figure out when the outlines are showing based on your actions. It seems it uses the same debounce as the movers, perhaps we could decouple them, or bring them both down to 100ms. (I also feel the movers remain open a bit too stubbornly.)
- Moving blocks is a bit rough now (the movers close as soon as you click on them and the initial click often doesn't trigger moving because they hide).
- Is there a way to simplify the "states" (is-highlighted, is-focused, etc) somehow or do we need them separate?

jasmussen commented:
This is what I see:

![highlight](https://user-images.githubusercontent.com/1204802/76850081-ebb92400-6846-11ea-9173-6f6f18fc5d77.gif)

The principle here feels very solid, and it is extra helpful when moving blocks:

![movers](https://user-images.githubusercontent.com/1204802/76850161-186d3b80-6847-11ea-965c-ee6954250928.gif)

I would echo Matías comments exactly, by the way. As you can see in the GIF above, the mover control also seems to collapse whenever a moving action is taken. The mover control should just remain visible whenever focus is inside. 

To Matías point about simplifying the states, it I wonder if you can take inspiration from his PR on adding a fade-effect between edit and select modes: https://github.com/WordPress/gutenberg/pull/20933 — that very same fade would be really nice to have here as well!

ItsJonQ commented:
Thanks for the feedback all!

> Let's try without the debounce for the highlight

@mtias I can give that a try

> the movers close as soon as you click on them and the initial click

I noticed this yesterday.

Reason: When the move happens, *something* in the code is forcing focus from the `button` to the `body` and back again. This brief focus shift messes with the show/hide logic.

 It's tricky. I'll figure something out.

> Is there a way to simplify the "states" (is-highlighted, is-focused, etc) somehow or do we need them separate

I believe we can simplify 💪 

> I wonder if you can take inspiration from his PR on adding a fade-effect between edit and select modes

@jasmussen I can look into the fade effect! Thanks for the suggestion 🤗

jasmussen commented:
> Reason: When the move happens, something in the code is forcing focus from the button to the body and back again. This brief focus shift messes with the show/hide logic.

Is this also the case in master?

Awesome work!

ItsJonQ commented:
> Is this also the case in master?

@jasmussen It is not (phew). I think the state updates to render the outline is making it unhappy.

Will work on refining 💪

Commit:
Fix janky Mover interactions when clicking

Resolved by correctly handling the unmount cycle for useEffect to reset the
highlight block state.

Commit:
Add transition animations for outlines

From https://github.com/WordPress/gutenberg/pull/20933

Commit:
Fix mouseLeave interaction by adjusting onChange callback

Commit:
Improve Mover/Highlight experience. Reduce debounce timeout from 500 -> 250

ItsJonQ commented:
@jasmussen + @mtias I pushed up changes that improve the experience as suggested (above).

If you folks would test it again, that would be lovely!

> Is there a way to simplify the "states" (is-highlighted, is-focused, etc) somehow or do we need them separate

I took a look. Although these states render the same UI (borders around the block, rendered with `:pseudo` elements), the meaning behind them are different.

| State               | Mode              | Description               | Visual      |
| ------------------- | ----------------- | ------------------------- | ----------- |
| `is-focused`        | Focus mode        | Current highlight         | `opacity: 1` |
| `is-selected`       | Navigation mode   | Current item              | `border`    |
| `is-multi-selected` | Multi-select mode | Selected item             | `border`    |
| `is-highlighted`    | Edit mode         | Toolbar interaction (New) | `border`    |

With that being said, I think keeping them separate is okay for now.
(Hopefully the table breakdown above helps!)

jasmussen commented:
This is what I see:

![show outline](https://user-images.githubusercontent.com/1204802/76937943-08606500-68f6-11ea-8c61-f1b3d5c753d8.gif)

This feels REALLY solid. I was about to suggest that you could reduce those 250ms to 0 for fading the outline out, but I realize it's very nice that the outline fades out at the same time as the mover control collapses. 

Overall this feels like a substantial improvement purely on behavior, and I'd love for us to ship this as fast as we can. In that vein, the only remaining discussion is the matrix of states. I do wish we could collapse some of those. It seems like the focus state might be leveraged with the "isolation" interface that FSE might enter when editing a template part. I also wonder if we can merge is-selected with is-multi-selected, with the toolbar differentiation happening elsewhere. But this is sort of high level and separate. 

If there isn't realy a way to attach the highlight without this extra state class, I'm personally fine with it. 

Nice work Q!

ItsJonQ commented:
> This feels REALLY solid. 

@jasmussen Thank you!!

> very nice that the outline fades out at the same time as the mover control collapses.

Awesome! Is there anything you think we need to adjust?

> If there isn't really a way to attach the highlight without this extra state class

Assuming all the various state `className` are purely for styles and have no semantic meaning (or things depending on them)... we could do something like this:

<img width="638" alt="Screen Shot 2020-03-18 at 9 51 18 AM" src="https://user-images.githubusercontent.com/2322354/76967630-0e5a4400-68fe-11ea-97ca-86253c6157e3.png">

Where a new singular `is-outlined` className is triggered by any of these conditions.

I think it would be out of scope for this PR. Just sharing for consideration :)

jasmussen commented:
I generally like the sound of that! Agreed probably out of scope for this one, but a good thought.

mtias commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1621,3 +1621,15 @@ export function isNavigationMode( state ) {
+ *
+ * @return {boolean} Whether the block is currently highlighted.
+ */
+export function getIsBlockHighlighted( state, clientId ) {
```
This should be named just `isBlockHighlighted` for consistency.

mtias commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1448,6 +1448,26 @@ export function automaticChangeStatus( state, action ) {
+ *
+ * @return {string} Updated state.
+ */
+export function blockHighlighted( state = '', action ) {
```
This one read a bit odd to me — perhaps `highlightedBlock` is better?

mtias reviewed (approved):
Works great for me now. Just left a couple comments on naming of functions that could be improved.

mtias commented:
@ItsJonQ actually I'm seeing one small thing — after moving a block, the highlighted state seems stuck. Initially it makes sense because focus moves to the movers, but if I focus on another element of the toolbar it still doesn't go away.

![image](https://user-images.githubusercontent.com/548849/76969915-5a6fad80-692b-11ea-89d5-263779316738.png)

ItsJonQ commented:
@mtias Awesome! Thank you for that feedback! I'll work on refining that as soon as I can 💪

Commit:
Update docs

ItsJonQ commented:
Why is Travis so mad at these updates 😭

Commit:
Rename highlightedBlock state and fix Toolbar blur interaction

aduth commented:
Couple quick questions:

- Could this not leverage the existing "is typing" flag, rather than specifically a separate new interaction of typing vs. not typing vs. interacting with the block toolbar?
- Can you confirm this behavior should apply when using "Top Toolbar" mode as well?
- Is a toolbar rendered within the context of a block such that we don't need a separate global state, but rather that we could share some state between the block and the toolbar to know that the block itself should be highlighted?

ItsJonQ commented:
@aduth Thank you for replying 🙏 

> Could this not leverage the existing "is typing" flag...

I don't think so, as it only engages when interacting specifically with these 2 elements:

![Screen Shot 2020-03-18 at 15 39 35](https://user-images.githubusercontent.com/2322354/77000409-bab41e80-692e-11ea-8c89-ea344c89204d.png)

> Can you confirm this behavior should apply when using "Top Toolbar" mode as well?

Yes. I believe it should.

> Is a toolbar rendered within the context of a block such that we don't need a separate global state

Ah! If there's a way to do that (that you know of), that would be lovely.
I don't necessarily think global state would be needed for this interaction.
I just wasn't sure if there was something else I could hook into.

aduth commented:
> > Could this not leverage the existing "is typing" flag...
> 
> I don't think so, as it only engages when interacting specifically with these 2 elements:

Ah, I guess then the question is more of a "should it", but I trust the judgment of y'all here.

> > Is a toolbar rendered within the context of a block such that we don't need a separate global state
> 
> Ah! If there's a way to do that (that you know of), that would be lovely.
> I don't necessarily think global state would be needed for this interaction.
> I just wasn't sure if there was something else I could hook into.

I had looked briefly at the code. It may be pretty straight-forward with the contextual form of the block toolbar, I'm not quite as sure with the "top toolbar" variant of it (which is rendered in the toolbar, and not within the block itself).

aduth commented:
Specifically, there is a context object which is provided by a block so that descendent components can make use of certain pieces relevant to the block's own state:

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-edit/context.js

ItsJonQ commented:
@aduth A parent component that contains both `BlockToolbar` and `Block` would `RootContainer`:
https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-list/root-container.js

Perhaps having some shared context there?

ItsJonQ commented:
Hmm! I'm not sure if we can use `/block-edit/context.js` as the `BlockToolbar` (most child of `BlockPopover`) is in another layer, not within the Context Provider.

The only shared context is this `BlockNodes.Provider`:
https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-list/root-container.js#L18

Which... I'm not too sure what it does, as the context value never changes 🙃

aduth commented on packages/block-editor/src/store/actions.js:
```diff
@@ -987,3 +987,21 @@ export function* insertAfterBlock( clientId ) {
 }
+
+/**
+ * Generator that toggles the highlighted block state.
```
This function isn't a generator, it's just a plain action creator.

More generally, I wonder if we shouldn't just standardize on some phrasing like "action dispatcher" or "action creator", considering:

- For developers, there's a high likelihood of copying documentation as prior art from other functions in this file, and we can optimize for this by standardizing on a consistent way to express them.
- For public documentation, there's really no need to differentiate between generators and action creators. Even if strictly speaking this thing we're exporting is either an action creator (plain function) or generator, for the purposes of documentation the only thing the reader should really care about is that this thing causes action(s) to be dispatched.

aduth commented on packages/block-editor/src/store/reducer.js:
```diff
@@ -1448,6 +1448,26 @@ export function automaticChangeStatus( state, action ) {
+			return action.clientId;
+
+		case 'UNHIGHLIGHT_BLOCK':
+			return '';
```
I think `null` would be a more meaningful data type to express the "unset" state. There's quite a bit of precedent as well (`block-editor`'s `insertionPoint` state, `editor`'s `postId` and `postType` states, etc).

aduth commented on packages/block-editor/src/store/actions.js:
```diff
@@ -987,3 +987,21 @@ export function* insertAfterBlock( clientId ) {
+	}
+	return {
+		type: 'UNHIGHLIGHT_BLOCK',
+	};
```
Two thoughts:

- Do we need separate action type for this, if we're already expressing the function as a single "toggle" operation? Typically I would expect the action creator to align to its own type, i.e. `TOGGLE_BLOCK_HIGHLIGHT`, or vice-versa, if we have separate types, then separate functions `highlightBlock`, `unhighlightBlock`. I have a strong preference toward the former (a single action type).
- Related to this point, I've been wondering if there's a possibility of race conditions where a block's unmounting (and subsequent "unhighlight" action) might occur only _after_ another block has opted to become "highlighted", and thus wrongly unhighlight the next highlighted block. Having a single action type might help with this, because if we have:

```js
return {
	type: 'TOGGLE_BLOCK_HIGHLIGHT',
	clientId,
	isHighlighted,
};
```

...then the reducer could verify that the "unhighlight" intent matches its own expectations for what the highlighted block is:

```js
function highlightedBlock( state = null, action ) {
	if ( action.type === 'TOGGLE_BLOCK_HIGHLIGHT' ) {
		if ( action.isHighlighted ) {
			return action.clientId;
		} else if ( state === action.clientId ) {
			return null;
		}
	}

	return state;
}
```

(My example could perhaps be expressed more elegantly)

aduth commented:
Could someone explain what is special about movers and block switching that they and only they should cause this effect?

mtias commented:
> Could someone explain what is special about movers and block switching that they and only they should cause this effect?

Those actions focus on the block as a whole and it's useful to see its boundaries (same as navigation mode).

ItsJonQ commented:
> Could someone explain what is special about movers and block switching that they and only they should cause this effect?

@aduth Sure!!

## Block Type

![Screen Shot 2020-03-18 at 16 37 46](https://user-images.githubusercontent.com/2322354/77005226-d28fa080-6936-11ea-8045-4d960cf7d2c7.png)

The [G2 Update](https://github.com/WordPress/gutenberg/issues/18667) added a new interaction pattern where the Movers would show/hide for certain interaction patterns. Mostly, interacting with the Block Type. By default, the Movers are hidden to reduce visual clutter.

The interaction mechanics were consolidated into a React hook that the `BlockToolbar` used.

## Outline

@mtias [had mentioned](https://github.com/WordPress/gutenberg/issues/20875) that the block should show some indication when the Block Type/Mover was interacted with.

The new "highlighted" state uses the same React hook mechanics (previously implemented for G2)

Visually, the styles are the same as `multi-select` or selection via Navigator mode (pressing `Escape`). However, the triggering mechanism is different.

Hope this helps!

aduth commented:
Thanks @mtias and @ItsJonQ . Where I may be getting caught up is that in many blocks, many of its toolbar buttons also apply to the block as a whole (list style, column vertical alignment, heading level, paragraph alignment). It wasn't previously clear to me that the mover and block type button were "linked" already, so at least I see the consistency in this approach.

Commit:
Improve removal of highlight state to work with editor events + rendering

Commit:
Fix inserter interaction by passing through events from .block-editor-block-list__block

Commit:
Rename getIsBlockHighlighted to isBlockHighlighted

Commit:
Fix commetn of toggleBlockHighlight action

Commit:
Simplify highlightedBlock reducer logic

ItsJonQ commented:
@aduth Thank you for your suggestions on the `TOGGLE_BLOCK_HIGHLIGHT` action/reducer! I just pushed an update with your proposed changes. 

> (My example could perhaps be expressed more elegantly)

I think you expressed it just fine 😊

aduth committed:
Docs: Update toggleBlockHighlight documentation

aduth reviewed (approved):
I left one comment worth considering. Otherwise, this is looking in good shape 👍

ItsJonQ commented on packages/block-editor/src/components/block-toolbar/utils.js:
```diff
@@ -38,10 +50,10 @@ export function useDebouncedShowMovers( {
 			}
 		},
-		[ showMovers ]
+		[ handleOnChange, showMovers ]
```
Thanks for your feedback!

In this case, the state of `showMovers` works as a guard for both `setShowMovers` and `onChange`. We only want to make the state change if it's different. 

Re: `handleOnChange`. Are you suggesting something like this?

```jsx
const handleOnChange = useCallback(( nextIsFocused ) => {
    setShowMovers( nextIsFocused );
    onChange( nextIsFocused );
}, [ onChange ])
```

And with this, I think we can remove it from being a dependency from the debounced Show/Hide functions

🤔

aduth replied:
Probably more the opposite, to instead remove `useCallback` from `debouncedShowMovers`. Over time, I've come to see `useCallback` as something of a code smell.

i.e.

```
function debouncedShowMovers( event ) {
	if ( event ) {
		event.stopPropagation();
	}

	const timeout = timeoutRef.current;

	if ( timeout && clearTimeout ) {
		clearTimeout( timeout );
	}
	if ( ! showMovers ) {
		handleOnChange( true );
	}
}
```

Would you see a problem with this alternative?

aduth replied:
> In this case, the state of `showMovers` works as a guard for both `setShowMovers` and `onChange`. We only want to make the state change if it's different.

But it's only incidentally accounting for changes in `onChange` on account for the fact that `handleOnChange` is a new function each render, which counteracts any benefits we might otherwise get from using `useCallback`. In other words, `debouncedShowMovers` will also be a new reference on every render.

ItsJonQ replied:
> Would you see a problem with this alternative? For what reason do we want useCallback ?

No problems 😊 . I was using `useCallback` as that was something I've seen in many places in our codebase, and it felt like a good idea to continue. But like you've suggested, it may be pre-optimizing, and perhaps causing more trouble and it's worth.

aduth replied:
The way I've understood it is that `useCallback` is useful if and only if there's an important reason to avoid changing references, and that in all other cases it's not only premature, but often yields _worse_ performance.

Related: https://kentcdodds.com/blog/usememo-and-usecallback

ItsJonQ replied:
KCD saves the day!

Commit:
Remove useCallback for debounce methods from useDebouncedShowMovers hook

aduth commented on packages/block-editor/src/components/block-toolbar/utils.js:
```diff
@@ -26,38 +38,32 @@ export function useDebouncedShowMovers( {
-		},
-		[ isFocused ]
-	);
+		timeoutRef.current = setTimeout( () => {
```
It existed previously, but: Is there any chance that this function is called twice before the callback is invoked? Because it could cause a problem where the `timeoutRef.current` is assigned to whichever `setTimeout` was scheduled last, and only the last one would be unscheduled in the `useEffect` unsubscribe callback (i.e. the first one would still trigger its callback even after the component unmounts, causing a warning to be logged if it calls to set state).

ItsJonQ replied:
Oh! Hmm... Perhaps we should clear the timeout before setting it in this debounce function.

aduth commented on packages/block-editor/src/components/block-toolbar/utils.js:
```diff
@@ -135,3 +145,33 @@ export function useShowMoversGestures( { ref, debounceTimeout = 500 } ) {
+		return () => {
+			// Sequences state change to enable editor updates (e.g. cursor
+			// position) to render correctly.
+			requestAnimationFrame( () => {
```
I don't have any other suggestions, but this code gives me vibes that it could be fragile (race conditions).

(`requestAnimationFrame` and `setTimeout` are often misused to address issues where we _should_ be acting as a side effect of some specific state transitions, and only be lucky circumstance can the same effect be achieved with scheduled callbacks)

ItsJonQ replied:
This was my only work around :(. I'm not comfortable with the fragile vibes either. There's something happening (elsewhere in the editor) that does something with focus (or perhaps caret position). Without `rAF` the interaction breaks.

I realized this after the E2E tests kept failing (earlier)

Open to suggestions! 🙏

youknowriad replied:
I personally wonder if we need this unmount behavior at all. It could be just a matter of including a reducer logic where the flag should be set to false when we start typing or when the block is unselected. (Unless I'm not understanding the purpose of this hook)

youknowriad replied:
I merged but if we find a better solution, let's follow-up on it.

Commit:
Improve and refactor clearTimeoutRef

ItsJonQ commented:
Rebased with latest `master`

mtias commented:
Thanks for getting this one in, it adds a very meaningful detail to the experience.

ItsJonQ commented:
Thank you all! 🙏 !
